add filter_time function

jf15558 opened the pull request:
# Function name
filter_time

## Description
A function to trim fossil occurrences to within a specified temporal duration and/or level of precision, using either stratigraphic interval names or absolute numeric values. There is an additional convenience feature, which I have had recent cause to use, to permit some occurrences which fall above a specified level of temporal precision to be retained for the sake of taxonomic completeness of the occurrence dataset.

Current commit contains commented code, documentation and examples for code review. testhat to be added following any changes from code review.

## Motivation and Context

Trimming and filtering by stratigraphic duration is a common preliminary phase of many palaeobiological work flows. While such a procedure can be readily implemented using logical subsetting in base R, there are multiple ways of achieving this, hence why a single standardised function will conceivably be of use.

## Checklist before requesting a review
- [x] I have read palaeoverse's [standards and structure](https://palaeoverse.org/articles/structure-and-standards.html) 
- [x] I have performed a self-review of my code.
- [x] I have added function documentation.
- [x] I have provided sufficient examples of implementation.
- [  ] If it is a core feature, I have added thorough tests.

jf15558 committed:
add filter_time function

A function to trim fossil occurrences to within a specified temporal
duration and/or level of precision. Current commit contains commented, code, documentation and examples for review. testhat to be added following any changes from code review

willgearty reviewed (changes requested):
Sorry for the delay on this review @jf15558, but this is a really great first draft! I will definitely be using this function a lot in my own research scripts!

I have a batch of comments, but it's definitely very close already. Don't forget you'll still need the following before it can be merged:

- generated docs (`devtools::document()`) (I wonder if we could automate this in the future...)
- add filter_time to an appropriate category in _pkgdown.yml
- add a `NEWS.md` item

jf15558 committed:
response to review - time filtering

All changes and additions proposed by Will have been implemented. testthat and documentation added. check() returned no issues

willgearty commented on R/filter_occs_time.R:
```diff
@@ -0,0 +1,277 @@
+#' filter_occs_time
```
Sorry, I wasn't saying you needed to change the name of the function (in fact, I think the old function name worked quite well and matches our existing naming scheme), just that this first line should literally be a title for the function (see suggestion)

```suggestion
#' Filter fossil occurrences by time
```

willgearty commented on R/filter_occs_time.R:
```diff
@@ -0,0 +1,277 @@
+#' @param max_age `numeric`. The maximum absolute occurrence age to trim to.
+#' Ignored if `oldest_bin` is supplied.
+#' @param uncertainty A numeric value denoting the maximum tolerable temporal
+#' uncertainty for retaining occurrences. Alternatively, if this argument is set
+#' to "bin", then only occurrences which span no more than `nbin` intervals in
+#' `bins` will be retained.
```
I definitely did not expect you to actually implement the `nbin` option, so maybe now this should be "bins" like before 🙈 (I had suggested "bin" because it was originally only one bin)? It probably doesn't matter (and you could also say "bins" but also catch "bin" as well), so I'll leave the decision up to you (and the other reviewer).

willgearty commented on R/filter_occs_time.R:
```diff
@@ -0,0 +1,277 @@
+#' @details
+#' When trimming by stratigraphic interval names, the boundary ages of
+#' `youngest_bin` and `oldest_bin` are taken from their entries in `bins`.
+#' Occurrences with maximum ages younger than or equal to the upper boundary
+#' age of `youngest_bin` or minimum ages older than or equal to the lower
+#' boundary age of `oldest_bin` will be discarded.
```
Unless I'm mistaken, I think this better matches what the code is doing here?
https://github.com/jf15558/palaeoverse/blob/b8df3dd76700c9f0532ae5c84c4c45dfe22145b3/R/filter_occs_time.R#L215-L216

```suggestion
#' Occurrences with maximum ages younger than the upper boundary
#' age of `youngest_bin` or minimum ages older than the lower
#' boundary age of `oldest_bin` will be discarded.
```

willgearty reviewed (changes requested):
Thanks for humoring my suggestions, @jf15558! Just a couple more comments. We'll also probably want a second review before merging, but I don't think I'll have any other comments after this.

jf15558 committed:
Second round of fixes

All suggestions implemented - function name reverted to filter_time(). In addition, nbin arg changed to n_bins to be in line with args in other palaeoverse functions

jf15558 committed:
More code coverage in tests

willgearty reviewed (approved):
Thanks again for addressing all of my little requests @jf15558! Looks like there might still be some edge cases that we could catch in tests ([see here](https://app.codecov.io/github/palaeoverse/palaeoverse/pull/142/blob/R/filter_time.R)), but it's not super necessary. Otherwise looks good to me!

Maybe @LewisAJones or @bethany-j-allen could contribute the second review?

jf15558 committed:
Modifications following second review

----- From BA comments -----

Documentation modified to indicate how the function can be used for occurrences with point ages (e.g. sampled from within age bounds).

Function updated to support inclusion or exclusion of occurrences which partly overlap with the set temporal bounds. Default behaviour is to exclude so that the returned data set falls strictly within the set bounds.

Function modified to report filtering progress to user (can be silenced).

----- other mods -----

Separate youngest_bin and min_ma arguments merged for simplicity (same for oldest_bin and max_ma). e.g., min_age can now be 411.2 or "Lochkovian".

Additional bug fixes to ensure the function behaves as described.

Documentation and consistency of phrasing improved.

jf15558 committed:
Update filter_time.R

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' Filter fossil occurrences by time
+#'
+#' @description
```
Currently, we don't actually include this roxygen tag in other function documentation (it's not strictly necessarily). However, perhaps we should start doing so... 

Any thoughts @willgearty @etiennebacher?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' intervals in `bins` will be retained.
+#' @param n_bins `numeric` The maximum tolerable bin-wise uncertainty above
+#' which occurrences will be discarded if `uncertainty = "bins"`.
+#' @param retain.taxa `logical`. Complete loss of some taxa may occur if all
```
Hmmm do we want to encourage this? I think it should be expected that we lose some data when filtering. With this argument, wouldn't we end up essentially keeping occurrences we are more uncertain about, while removing many occurrences which are more constrained? What does this mean for downstream analyses? 

Also, is there a strong reason to use "retain.taxa" over "retain_taxa"?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' occurrences with point or uncertainty-bounded ages using stratigraphic
+#' interval names or absolute numeric ages.
+#'
+#' @param occdf `data.frame`. The fossil occurrences you wish to filter. This
```
Obviously, we're going to be doing a big overhaul of the package docs including parameter definitions. However, I'd suggest reviewing other functions in the package and bringing this more in line with the wording elsewhere, e.g.
```suggestion
#' @param occdf `data.frame`. A dataframe of the fossil occurrences you wish to filter. This
```

LewisAJones replied:
This doesn't really matter, but just aiming for consistency with other function docs
```suggestion
#' @param occdf \code{data.frame}. The fossil occurrences you wish to filter. This
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' @param occdf `data.frame`. The fossil occurrences you wish to filter. This
+#' must contain minimum and maximum occurrence age columns, or a column of
+#' point ages. All must be numeric and cannot contain missing values.
+#' @param bins This may be left as `NULL` if filtering by numeric values only.
```
```suggestion
#' @param bins \code{data.frame}. This may be left as `NULL` if filtering by numeric values only.
```

I think it would also be best to flip this parameter definition. For example: 

A dataframe of time bins used to look up numeric ages for filtering fossil occurrences, such as that returned by [time_bins()]. This must contain at least three named columns: `max_ma`, `min_ma`, and `interval_name`. No column can contain missing values and columns `max_ma` and `min_ma` must be numeric. If `NULL` (default), filtering is by numeric values only.

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' `interval_name`. No column can contain missing values and columns `max_ma`
+#' and `min_ma` must additionally be numeric.
+#' @param min_ma `character`. The name of the column you wish to be treated as
+#' minimum ages in `occdf` and `bins`. `min_ma` in `occdf` can instead be
```
In the `bins` parameter documentation you state that the column must be named as "max_ma" and "min_ma", but here you suggest that it can be specified?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' @param min_age The minimum desired occurrence age. This can be either a
+#' numeric value or an interval name, which then must be present in column
+#' `interval_name` in `bins`.
+#' @param max_age The maximum desired occurrence age. This can be either a
```
As the function can take both ages and interval names (e.g. stages, periods, etc)... Is there a more generic argument name we can have here?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' treated as point ages.
+#' @param max_ma `character`. The name of the column you wish to be treated as
+#' maximum ages in `occdf` and `bins`. `max_ma` in `occdf` can instead be
+#' treated as point ages (i.e. the same column name as for `min_ma`).
```
We used to explicitly state we allowed point age estimates in `bin_time` but removed it to avoid confusion. I wonder if we should just mention how users could use the function with point age estimates in the details and provide an example?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' @param max_ma `character`. The name of the column you wish to be treated as
+#' maximum ages in `occdf` and `bins`. `max_ma` in `occdf` can instead be
+#' treated as point ages (i.e. the same column name as for `min_ma`).
+#' @param interval_name `character` The name of the column in `bins` that
```
```suggestion
#' @param interval_name `character`. The name of the column in `bins` that
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' @param max_age The maximum desired occurrence age. This can be either a
+#' numeric value or an interval name, which then must be present in column
+#' `interval_name` in `bins`.
+#' @param overlap `logical`. Occurrences which partially overlap with the set
```
Should this argument name be more explicit? e.g. keep_overlap/drop_overlap?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' filtering duration are discarded by default, so that the returned data falls
+#' strictly within these bounds. Set to `TRUE` to retain overlaps instead.
+#' @param uncertainty A numeric value denoting the maximum tolerable temporal
+#' uncertainty for retaining occurrences. Alternatively, if this argument is
```
Are there instances where individuals may want to combine both `uncertainty` and `n_bins`? In general, I think it is better when arguments don't interact too much... `n_bins` could just be applied when it isn't `NULL`? 

You could also have age_uncertainty and bin_uncertainty if you wanted to be really specific?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' their occurrences exceed the temporal uncertainty criterion. If this is
+#' undesirable, occurrences for these taxa will be retained as exceptions to
+#' the this filtering criterion.
+#' @param taxa `character`. The name of the column you wish to be treated as
```
If we removed `retain.taxa` this could also go, helping to streamline the function a little

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' the this filtering criterion.
+#' @param taxa `character`. The name of the column you wish to be treated as
+#' taxon names in `occdf` if `retain.taxa = TRUE`.
+#' @param verbose `logical`. Sets if the impact of time filtering is reported
```
I think some users would like this report, and Bethany seems to think so too, but personally I don't like the console being flooded with messages. If you want to keep this, I'd suggest having this argument set to FALSE by default. I'd also make the messaging more generic (e.g. X occurrences dropped). This might be something we want to eventually propagate across other functionality?... I can imagine a suite of `filter_*` functions

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' When filtering by stratigraphic bounds, minimum or maximum ages can be used
+#' individually, or at the same time. By default occurrences which do not fall
+#' fully within the bounds set by `min_age` and/or `max_age` are discarded.
+#' This behaviour can be toggled by setting `strict = FALSE` to retain
```
This argument doesn't exist?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' the interval durations in `bins`. Otherwise, the numeric minimum and maximum
+#' values are used directly.
+#'
+#' When filtering by stratigraphic precision, if `uncertainty = "bins"`, the
```
This is quite a high-level interaction between functions, and `bin_time` is likely to go through some changes in the coming months. This depends a little on how we want users to use our functions, but I could imagine this being a good opportunity for showing how things can be piped together instead? However, I do like the simplicity for users of including this here. I think this is a discussion point for us on this function.

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' or reasonable to use poorer quality data analytically.
+#'
+#' @section Developer(s): Joseph T. Flannery-Sutherland
+#' @section Reviewer(s): William Gearty, Bethany Allen, Lewis Jones
```
```suggestion
#' @section Reviewer(s): William Gearty, Bethany Allen, Lewis A. Jones
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' @section Reviewer(s): William Gearty, Bethany Allen, Lewis Jones
+#'
+#' @examples
+#' # get internal tetrapod data and GTS2020 stages
```
Standardising to other function examples docs (implement for all)
```suggestion
#' # Get internal tetrapod data and GTS2020 stages
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#' occdf <- tetrapods[1:100, ]
+#' stages <- time_bins(rank = "stage")
+#'
+#' # trim to Kungurian occurrences
```
I think we should avoid using "trim" and maybe stick to filter, remove, or exclude

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+                        min_ma = "min_ma", max_ma = "max_ma", interval_name = "interval_name",
+                        min_age = NULL, max_age = NULL, overlap = FALSE,
+                        uncertainty = NULL, n_bins = 1,
+                        retain.taxa = FALSE, taxa = NULL, verbose = T) {
```
```suggestion
                        retain.taxa = FALSE, taxa = NULL, verbose = FALSE) {
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+                        uncertainty = NULL, n_bins = 1,
+                        retain.taxa = FALSE, taxa = NULL, verbose = T) {
+
+  # occdf <- data.frame(genus   = c("A", "A", "B", "B", "B", "C", "D", "E", "E", "F"),
```
A reminder to remove this for function testing!

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+
+  # check occurrence data and copy for filtering
+  if (is.data.frame(occdf) == FALSE) {
+    stop("`occdf` should be a data.frame.")
```
This needs changing elsewhere in palaeoverse but I guess "should" implies some flexibility (which is not correct)
```suggestion
    stop("`occdf` must be a data.frame.")
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  # check arguments
+  if (any(c(!is.atomic(min_ma), length(min_ma) != 1, !min_ma %in% colnames(occdf),
```
Should this be `is.character`? 
```suggestion
  if (any(c(!is.character(min_ma), length(min_ma) != 1, !min_ma %in% colnames(occdf),
```

I would also say you're checking multiple conditions here and the error message doesn't necessarily help for some cases. For example, imagine a user makes a typo in the column name.

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+      stop("If not NULL, please specify `max_ma` as a column name in `occdf`.")
+    }
+  }
+  if (any(is.na(occdf[, min_ma])) || any(is.na(occdf[, max_ma]))) {
```
Could also be split out to help the user locate issues in their dataset?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+  if (any(is.na(occdf[, min_ma])) || any(is.na(occdf[, max_ma]))) {
+    stop(paste0("NA values detected in ", min_ma, " or ", max_ma, " in `occdf`"))
+  }
+  if (!is.numeric(occdf[, min_ma]) || !is.numeric(occdf[, max_ma])) {
```
Could also be split out to help the user locate issues in their dataset?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+      stop("Columns `min_ma` and `max_ma` in `bins` should contain numeric values.")
+    }
+    if (!any(c(bins[, max_ma], bins[, min_ma]) %in% c(occdf[, max_ma], occdf[, min_ma]))) {
+      warning("No shared ages between `bins` and `occdf`. Check the time scale and range of your data.")
```
Do they need to be shared?

I imagine we would want to check whether they are in range of each other and overlap (i.e. test maximum and minimum ages) but not whether exact numeric ages match?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+#'
+#' @export
+
+filter_time <- function(occdf, bins = NULL,
```
Is there value in having this split into a couple of functions? Something along the lines of:

- filter_by_age
- filter_by_interval

This could clean up and streamline a lot of the documentation?

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+    }
+    max_age <- bins[which(bins[, interval_name] == max_age), max_ma]
+  }
+  if (is.numeric(min_age)) {
```
I guess we should have already checked whether min_age / max_age are character or numeric in the error handling, so we could probably just use if and else statements for L234--258? Equally, this is also fine!

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+    }
+  }
+  if (is.numeric(max_age)) {
+    occdf <- occdf[which(occdf[,min_ma] < max_age),]
```
Small edit for code style/consistency (check throughout)
```suggestion
    occdf <- occdf[which(occdf[ ,min_ma] < max_age),]
```

LewisAJones commented on R/filter_time.R:
```diff
@@ -0,0 +1,340 @@
+  }
+  if (all(!is.null(min_age), !is.null(max_age))) {
+    if (min_age >= max_age) {
+      stop("The minimum filtering age is not younger than than the maximum filtering age.")
```
```suggestion
      stop("The minimum filtering age is greater than the maximum filtering age.")
```

LewisAJones reviewed (changes requested):
Hi @jf15558, sorry for the delay on reviewing this. I've now given this a once over, and I really like the intended purpose. 

As you'll see, I have a few comments, but my main point for discussion is whether this function should actually be split out into a couple of functions? Currently, you can filter using numeric ages or interval names. I like having these options but I wonder if it makes more sense to have a filter_by_age and filter_by_interval (or similar) function. This would streamline a lot of the documentation and the internal code making both easier to maintain and support the user in getting started quickly.

Minor points:
- I came across a few instances of typos, or outdated documentation, so it's probably worth doing a proofread before the next review stage.
- There are some cases where error handling and messaging could be clearer for the user (e.g. multiple error handling tests but the error message wouldn't always help the user in debugging).
- There are a few instances where code style does not strictly follow our standards ([tidyverse style guide](https://style.tidyverse.org)). We still need to improve some of this in the broader package, but better to catch this now!

I'll give this another, more thorough review once we've decided on the ultimate direction of this function. I don't think too fine edits are helpful at this stage!
